sklearn/utils/extmath.py

```diff
@@ -255,37 +257,71 @@ def randomized_range_finder(
     analysis
     A. Szlam et al. 2014
     """
+    xp, is_array_api_compliant = get_namespace(A)
     random_state = check_random_state(random_state)
 
     # Generating normal random vectors with shape: (A.shape[1], size)
-    Q = random_state.normal(size=(A.shape[1], size))
-    if hasattr(A, "dtype") and A.dtype.kind == "f":
-        # Ensure f32 is preserved as f32
-        Q = Q.astype(A.dtype, copy=False)
+    # XXX: generate random number directly from xp if it's possible
+    # one day.
+    Q = xp.asarray(random_state.normal(size=(A.shape[1], size)))
+    if hasattr(A, "dtype") and xp.isdtype(A.dtype, kind="real floating"):
+        # Use float32 computation and components if A has a float32 dtype.
+        Q = xp.astype(Q, A.dtype, copy=False)
+
+    # Move Q to device if needed only after converting to float32 if needed to
+    # avoid allocating unnecessary memory on the device.
+
+    # Note: we cannot combine the astype and to_device operations in one go
+    # using xp.asarray(..., dtype=dtype, device=device) because downcasting
+    # from float64 to float32 in asarray might not always be accepted as only
+    # casts following type promotion rules are guarateed to work.
+    # https://github.com/data-apis/array-api/issues/647
+    if is_array_api_compliant:
+        Q = to_device(Q, getattr(A, "device", None))
```

Comment:
At this point `Q` has the correct dtype, can we use `asarray` to do the device transfer?

```suggestion
    Q = xp.asarray(Q, device=device(A))
```

> Note that array_api_compat provides a device helper function that returns "cpu" for numpy arrays and array.device otherwise. We could do something similar instead of using getattr(array, "device", None) if you prefer.

We already have it in `device` defined in `sklearn.utils._array_api`.